source/ini_file.cpp

```diff
@@ -11,7 +11,7 @@
 #include <shared_mutex>
 
 static std::shared_mutex s_ini_cache_mutex;
-static std::unordered_map<std::wstring, std::unique_ptr<ini_file>> s_ini_cache;
+static std::unordered_map<std::wstring, std::shared_ptr<ini_file>> s_ini_cache;
```

Comment:
Simply keep this a `unique_ptr` and only ever add the canonical representation of a path the map (don't add multiple paths pointing to the same `ini_file` object). That will incur an additional lookup cost if the path passed to `load_cache`/`flush_cache` is not already canonical, since it first has to be transformed to that, but simplifies this cache.

Reply:
I actually don't know about the behavior of unique and shared pointers, and I'm not sure what problem will be solved by these classes.
If the total costs of the syscalls by running `canonical()` in usually is so small, then I think it is fine your proposal.